proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1038 @@
+  of ``X``.
+
+* Define the *strip-qualifier* operation on an environment ``E``::
+      
+      ∂(modid) E = { x ↦ orig | (modid.x ↦ orig) ∈ E }
+
+  A ``∂(modid) E`` is an exports: it maps *unqualified* names (only) to original names.
+  In English: ``∂(modid) E`` selects all these entities whose
+  qualified names starts with ``modid``, and remove the 
+  ``modid`` from their names (yielding a plain occurrence name).
+
+  Note that ``modid . ∂(modid) E`` is simply a filter, keeping all those names
+  in ``E`` which start with ``modid``.
+  
+* **Import specifiers**: An import specification, or *impspec*, can be listed
+  in parentheses in an ``import`` statement to restrict what is imported.
+  We say we *interpret* an *impspec* with respect to an export environment
+  to create an environment.
+  (An *impspec* may be specified
+  via complement, using ``hiding``; this proposal will not worry about this
+  detail.)
+
+* **Import item**:  An element in an import specifier is an *import item*.
+  Each import item is interpreted by looking it up in the export environment of
+  the module being imported. If this lookup fails, then report an error.
+  
+* **Imports**: An ``import`` statement names a module *M* and (optionally) an 
+  import-specifier *impspec*.
+
+  Let *X* be the export environment of *M*.
+  
+  Let *I* be the interpretation of *impspec* with respect to *X*. If
+  *impspec* is omitted, let *I = X*.
+  
+  Let the *module alias A* be the module name appearing after the ``as`` in
+  the ``import`` statement. If there is no ``as``, then the module alias *A*
+  is just *M*.
+
+  The ``import`` statement brings names into scope by adding entries
+  to the global environment. Concretely, add *A . I* to the environment.
+  If the ``import`` statement does not use
+  ``qualified``, additionally add *I* to the global environment.
+
+Core Proposal Change Specification
+----------------------------------
+
+This proposal's specification is divided into pieces. This core piece is a necessary
+component of the overall proposal. Later pieces can be chosen piecemeal.
+
+1. **Definitions**
+
+   A. A qualified name can now have any number of module qualifications (instead
+      of just 0 or 1), each
+      separated by a ``.`` and with no intervening whitespace. Formally, a qualified
+      name is the pair of a list (perhaps empty) of module names and an occurrence name.
+
+      The concrete can lead
+      to potential ambiguity: is the name ``Data.List.length`` qualified by
+      ``Data.List`` (a well-formed module name) or qualified both by ``Data``
+      and by ``List``? If ``length`` is in scope qualified both by ``Data.List``
+      and by both ``Data`` and ``List``, the occurrence ``Data.List.length`` is
+      an error. There is no way to disambiguate locally without using module
+      aliases or other renamings. (Programmers should seek to avoid this scenario.)
```

Comment:
This is actually a change. See the last [Effect and Interaction](https://github.com/goldfirere/ghc-proposals/blob/local-modules/proposals/0000-local-modules.rst#6effect-and-interactions). One more example:

```hs
module M ( module qualified Data ) where
import qualified Data.Set
import qualified Data.Map
```

This `M` would be in error, because there are no identifiers qualified by `Data`.

I suppose we could change this, though the change would be backward-incompatible (in case a module like the one above also imported a module `Data`).

Reply:
I don't really follow. Crucially, how is Haskell supposed to segment `X.Y.Z` into a list of segments?

For the example, `module qualified` is a new export item, so there is no change of backward incompatibility. Maybe there is a difference when using `module Data` as the export, but I'm not sure that I see it. Could you spell it out? I guess it would be exporting more stuff. While it is technically not backward compatible, it is not the sort of issue that are worth bending over backward for (for instance, pvp doesn't consider exporting more items to be worth a major version).